scalardb/src/scalardb/core.clj

```diff
@@ -123,38 +146,63 @@
                (reset! (:transaction test)))
           (catch Exception e
             (warn (.getMessage e))))
-        (when-not (nil? (:transaction test))
+        (when-not @(:transaction test)
           (recur (dec tries)))))))
 
+(defn check-connection!
+  [test]
+  (when (nil? @(:transaction test))
+    (prepare-transaction-service! test)))
+
+(defn try-reconnection!
+  [test]
+  (when (= (swap! (:failures test) inc) NUM_FAILURES_FOR_RECONNECTION)
+    (prepare-transaction-service! test)
+    (reset! (:failures test) 0)))
+
 (defn start-transaction
   [test]
   (some-> test :transaction deref .start))
 
+(defmacro with-retry
```

Comment:
Just a question. What is the point of using `defmacro` here instead of `defn` ? for performance reason ?

Reply:
I want to use macro `with-retry` for other functions which need to be retried when it fails so that we can deduplicate the retry code.
Each retry process is almost the same, but process functions and reconnection functions and arguments are different.